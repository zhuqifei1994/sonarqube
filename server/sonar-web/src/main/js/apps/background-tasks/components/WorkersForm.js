/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import Modal from 'react-modal';
import Select from 'react-select';
import { times } from 'lodash';
import { setWorkerCount } from '../../../api/ce';
import { translate } from '../../../helpers/l10n';

const MAX_WORKERS = 10;

type Props = {
  onClose: (newWorkerCount?: number) => void,
  workerCount: number
};

type State = {
  newWorkerCount: number,
  submitting: boolean
};

export default class WorkersForm extends React.PureComponent {
  mounted: boolean;
  props: Props;
  state: State;

  constructor(props: Props) {
    super(props);
    this.state = {
      newWorkerCount: props.workerCount,
      submitting: false
    };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleClose = () => this.props.onClose();

  handleWorkerCountChange = (option: { value: number }) =>
    this.setState({ newWorkerCount: option.value });

  handleSubmit = (event: Event) => {
    event.preventDefault();
    this.setState({ submitting: true });
    const { newWorkerCount } = this.state;
    setWorkerCount(newWorkerCount).then(
      () => {
        if (this.mounted) {
          this.props.onClose(newWorkerCount);
        }
      },
      () => {
        if (this.mounted) {
          this.setState({ submitting: false });
        }
      }
    );
  };

  render() {
    const options = times(MAX_WORKERS).map((_, i) => ({ label: i + 1, value: i + 1 }));

    return (
      <Modal
        isOpen={true}
        contentLabel={translate('background_tasks.change_number_of_workers')}
        className="modal"
        overlayClassName="modal-overlay"
        onRequestClose={this.handleClose}>
        <header className="modal-head">
          <h2>{translate('background_tasks.change_number_of_workers')}</h2>
        </header>
        <form onSubmit={this.handleSubmit}>
          <div className="modal-body">
            <Select
              className="input-tiny spacer-top"
              clearable={false}
              onChange={this.handleWorkerCountChange}
              options={options}
              searchable={false}
              value={this.state.newWorkerCount}
            />
            <div className="big-spacer-top alert alert-success markdown">
              {translate('background_tasks.change_number_of_workers.hint')}
            </div>
          </div>
          <footer className="modal-foot">
            <div>
              {this.state.submitting && <i className="spinner spacer-right" />}
              <button disabled={this.state.submitting} type="submit">
                {translate('save')}
              </button>
              <button type="reset" className="button-link" onClick={this.handleClose}>
                {translate('cancel')}
              </button>
            </div>
          </footer>
        </form>
      </Modal>
    );
  }
}
