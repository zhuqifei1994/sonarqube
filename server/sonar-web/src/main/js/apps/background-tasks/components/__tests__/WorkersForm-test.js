/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import { shallow } from 'enzyme';
import WorkersForm from '../WorkersForm';
import { submit, doAsync } from '../../../../helpers/testUtils';

jest.mock('../../../../api/ce', () => ({
  setWorkerCount: () => Promise.resolve()
}));

it('changes select', () => {
  const wrapper = shallow(<WorkersForm onClose={jest.fn()} workerCount={1} />);
  expect(wrapper).toMatchSnapshot();

  wrapper.find('Select').prop('onChange')({ value: 7 });
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it('returns new worker count', () => {
  const onClose = jest.fn();
  const wrapper = shallow(<WorkersForm onClose={onClose} workerCount={1} />);
  // $FlowFixMe
  wrapper.instance().mounted = true;
  wrapper.find('Select').prop('onChange')({ value: 7 });

  wrapper.update();
  submit(wrapper.find('form'));

  return doAsync(() => {
    expect(onClose).toBeCalled();
  });
});
