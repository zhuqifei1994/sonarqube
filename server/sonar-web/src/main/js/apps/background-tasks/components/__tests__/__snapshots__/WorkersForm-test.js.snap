// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes select 1`] = `
<Modal
  ariaHideApp={true}
  className="modal"
  closeTimeoutMS={0}
  contentLabel="background_tasks.change_number_of_workers"
  isOpen={true}
  onRequestClose={[Function]}
  overlayClassName="modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <header
    className="modal-head"
  >
    <h2>
      background_tasks.change_number_of_workers
    </h2>
  </header>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal-body"
    >
      <Select
        addLabelText="Add \\"{label}\\"?"
        arrowRenderer={[Function]}
        autosize={true}
        backspaceRemoves={true}
        backspaceToRemoveMessage="Press backspace to remove {label}"
        className="input-tiny spacer-top"
        clearAllText="Clear all"
        clearValueText="Clear value"
        clearable={false}
        delimiter=","
        disabled={false}
        escapeClearsValue={true}
        filterOptions={[Function]}
        ignoreAccents={true}
        ignoreCase={true}
        inputProps={Object {}}
        isLoading={false}
        joinValues={false}
        labelKey="label"
        matchPos="any"
        matchProp="any"
        menuBuffer={0}
        menuRenderer={[Function]}
        multi={false}
        noResultsText="No results found"
        onBlurResetsInput={true}
        onChange={[Function]}
        onCloseResetsInput={true}
        openAfterFocus={false}
        optionComponent={[Function]}
        options={
          Array [
            Object {
              "label": 1,
              "value": 1,
            },
            Object {
              "label": 2,
              "value": 2,
            },
            Object {
              "label": 3,
              "value": 3,
            },
            Object {
              "label": 4,
              "value": 4,
            },
            Object {
              "label": 5,
              "value": 5,
            },
            Object {
              "label": 6,
              "value": 6,
            },
            Object {
              "label": 7,
              "value": 7,
            },
            Object {
              "label": 8,
              "value": 8,
            },
            Object {
              "label": 9,
              "value": 9,
            },
            Object {
              "label": 10,
              "value": 10,
            },
          ]
        }
        pageSize={5}
        placeholder="Select..."
        required={false}
        scrollMenuIntoView={true}
        searchable={false}
        simpleValue={false}
        tabSelectsValue={true}
        value={1}
        valueComponent={[Function]}
        valueKey="value"
      />
      <div
        className="big-spacer-top alert alert-success markdown"
      >
        background_tasks.change_number_of_workers.hint
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <div>
        <button
          disabled={false}
          type="submit"
        >
          save
        </button>
        <button
          className="button-link"
          onClick={[Function]}
          type="reset"
        >
          cancel
        </button>
      </div>
    </footer>
  </form>
</Modal>
`;

exports[`changes select 2`] = `
<Modal
  ariaHideApp={true}
  className="modal"
  closeTimeoutMS={0}
  contentLabel="background_tasks.change_number_of_workers"
  isOpen={true}
  onRequestClose={[Function]}
  overlayClassName="modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <header
    className="modal-head"
  >
    <h2>
      background_tasks.change_number_of_workers
    </h2>
  </header>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal-body"
    >
      <Select
        addLabelText="Add \\"{label}\\"?"
        arrowRenderer={[Function]}
        autosize={true}
        backspaceRemoves={true}
        backspaceToRemoveMessage="Press backspace to remove {label}"
        className="input-tiny spacer-top"
        clearAllText="Clear all"
        clearValueText="Clear value"
        clearable={false}
        delimiter=","
        disabled={false}
        escapeClearsValue={true}
        filterOptions={[Function]}
        ignoreAccents={true}
        ignoreCase={true}
        inputProps={Object {}}
        isLoading={false}
        joinValues={false}
        labelKey="label"
        matchPos="any"
        matchProp="any"
        menuBuffer={0}
        menuRenderer={[Function]}
        multi={false}
        noResultsText="No results found"
        onBlurResetsInput={true}
        onChange={[Function]}
        onCloseResetsInput={true}
        openAfterFocus={false}
        optionComponent={[Function]}
        options={
          Array [
            Object {
              "label": 1,
              "value": 1,
            },
            Object {
              "label": 2,
              "value": 2,
            },
            Object {
              "label": 3,
              "value": 3,
            },
            Object {
              "label": 4,
              "value": 4,
            },
            Object {
              "label": 5,
              "value": 5,
            },
            Object {
              "label": 6,
              "value": 6,
            },
            Object {
              "label": 7,
              "value": 7,
            },
            Object {
              "label": 8,
              "value": 8,
            },
            Object {
              "label": 9,
              "value": 9,
            },
            Object {
              "label": 10,
              "value": 10,
            },
          ]
        }
        pageSize={5}
        placeholder="Select..."
        required={false}
        scrollMenuIntoView={true}
        searchable={false}
        simpleValue={false}
        tabSelectsValue={true}
        value={7}
        valueComponent={[Function]}
        valueKey="value"
      />
      <div
        className="big-spacer-top alert alert-success markdown"
      >
        background_tasks.change_number_of_workers.hint
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <div>
        <button
          disabled={false}
          type="submit"
        >
          save
        </button>
        <button
          className="button-link"
          onClick={[Function]}
          type="reset"
        >
          cancel
        </button>
      </div>
    </footer>
  </form>
</Modal>
`;
